Outer Tracker Phase2 DQM: additional stub validation plots

brandiskip opened the pull request:
#### PR description:

Histograms added for stub coordinates Δφ and Δz, stub Δbend, and stub efficiency vs tp pT
[Proposal for validation plots](https://indico.cern.ch/event/1242120/contributions/5230280/attachments/2579402/4449460/DQM_Tracker.pdf)
[Presentation slides with final results](https://indico.cern.ch/event/1414546/contributions/5965160/attachments/2861743/5006999/DQM_Update_OfflineSW.pdf)

#### PR validation:

Added histograms examined and validated, showing sensible results as presented at the offline software meeting

ran:
scram build code-checks
scram build code-format

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"0722806432fe0edc0fa7aa246e67e72078c18b4e":{"files":["Validation/SiTrackerPhase2V/python/Phase2OTValidateTrackingParticles_cfi.py"],"squashed":false,"time":1721138497},"19177d3d93a99e54eead8c047edd40c375050e17":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py","DQM/SiTrackerPhase2/test/harvestingstep_phase2tk_cfg.py","Validation/SiTrackerPhase2V/python/Phase2OTValidateTrackingParticles_cfi.py","Validation/SiTrackerPhase2V/python/Phase2TrackerValidationFirstStep_cff.py"],"squashed":false,"time":1717678465},"25d9a1cbc9a69652fed4e16900f78955e7a4b4ee":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py"],"squashed":false,"time":1721138307},"3e6ad0e83050bc23f1a8f2379f2a724a0d3db64b":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py","Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestTrackingParticles.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc","Validation/SiTrackerPhase2V/python/Phase2OTValidateTrackingParticles_cfi.py","Validation/SiTrackerPhase2V/python/Phase2TrackerValidationFirstStep_cff.py"],"squashed":false,"time":1718891281},"419c3b2546f40955176d9fcb71a08afd52f44b01":{"files":["Validation/SiTrackerPhase2V/python/Phase2OTValidateTTStub_cfi.py"],"squashed":false,"time":1719919914},"429a00a7f9b4f33e18b59b0e5f2b5eb925671593":{"files":["Validation/SiTrackerPhase2V/python/Phase2OTValidateTTStub_cfi.py"],"squashed":false,"time":1718972296},"4546abb9822c0521f9ecaf4cfa9675d003b6ce8d":{"files":["Validation/SiTrackerPhase2V/python/Phase2OTValidateTTStub_cfi.py"],"squashed":false,"time":1718971820},"4ab0c8b403c1928b4a8f20951a76a4741c3d4ebe":{"files":["DQM/SiTrackerPhase2/test/DQM_V0001_R000000001__Global__CMSSW_X_Y_Z__RECO.root","DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py","DQM/SiTrackerPhase2/test/step3_pre4_inDQM.root","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1718974032},"525f1d5c2501a904a9003a3834d04d1e39a2925b":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py"],"squashed":false,"time":1721159815},"65e249282f6873c8d6b924329bd2a9235a16bf08":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1719919863},"861c78328c8a94ad1b956557a0ff81c6466e45bd":{"files":["Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc"],"squashed":false,"time":1718979787},"98f89e8f0404fdda07bafbeb96d25e98384033ad":{"files":["DQM/SiTrackerPhase2/test/DQM_V0001_R000000001__Global__CMSSW_X_Y_Z__RECO.root","DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py","DQM/SiTrackerPhase2/test/step3_pre4_inDQM.root","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc","Validation/SiTrackerPhase2V/python/Phase2OTValidateTTStub_cfi.py"],"squashed":false,"time":1719841331},"a44deb4fbbfd32880a9cf600b348b513497e7175":{"files":["DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py","Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestTrackingParticles.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc","Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc","Validation/SiTrackerPhase2V/python/Phase2OTValidateTrackingParticles_cfi.py"],"squashed":false,"time":1721052995},"fbfdfd34542476581ff537234fac4bc6cda48487":{"files":[],"squashed":false,"time":1718975697},"fc748d539055a605f808438c03c5ca9fde34ba7a":{"files":["Validation/SiTrackerPhase2V/python/Phase2OTValidateTTStub_cfi.py"],"squashed":false,"time":1718921797}},"emoji":{},"last_seen_sha":"525f1d5c2501a904a9003a3834d04d1e39a2925b","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/40939

- Found files with invalid states:
  - DQM/SiTrackerPhase2/test/step3_pre4_inDQM.root:
    - Added: 4ab0c8b403c1928b4a8f20951a76a4741c3d4ebe
    - Deleted: 98f89e8f0404fdda07bafbeb96d25e98384033ad
  - DQM/SiTrackerPhase2/test/DQM_V0001_R000000001__Global__CMSSW_X_Y_Z__RECO.root:
    - Added: 4ab0c8b403c1928b4a8f20951a76a4741c3d4ebe
    - Deleted: 98f89e8f0404fdda07bafbeb96d25e98384033ad

cmsbuild commented:
A new Pull Request was created by  @brandiskip for master.

It involves the following packages:

- DQM/SiTrackerPhase2 (**dqm**)
- Validation/SiTrackerPhase2V (**dqm**)


@antoniovagnerini, @cmsbuild, @nothingface0, @rvenditti, @syuvivida, @tjavaid can you please review it and eventually sign? Thanks.
@arossi83, @missirol, @mmusich, @sroychow this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

sarafiorendi commented on DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py:
```diff
@@ -1,22 +1,22 @@
-import FWCore.ParameterSet.Config as cms
+# with command line options: step3 --conditions auto:phase2_realistic_T21 -s RAW2DIGI,L1Reco,RECO,RECOSIM,VALIDATION:@phase2Validation,DQM:@phase2 --datatier DQMIO -n 10 --geometry Extended2026D98 --era Phase2C11M9 --eventcontent DQM --no_exec
 
+import FWCore.ParameterSet.Config as cms
```
test comment from sara

tomalin commented:
In https://indico.cern.ch/event/1414546/contributions/5965160/attachments/2861743/5006999/DQM_Update_OfflineSW.pdf 
1) Slide 4 shows stub resolution in z. Is this really the resolution in r-z? The modules measure z in the barrel, but r in the endcap.
2) Slide 7 gives no info about the stub efficiency for Pt < 2 GeV. There is probably a little efficiency in the range between 1.5 - 2 GeV, so it would be nice to see it.

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:
```diff
@@ -38,96 +44,475 @@
+
+  // delta_phi hists
+  MonitorElement* phi_res_barrel;
+  MonitorElement* phi_res_barrel_L1;
```
Instead of declaring separately phi_res_barrel_L1, phi_res_barrel_L2 etc., which requires many lines of code, use an std::vector<MonitorElement*>. Then when booking and filling the histograms, you can use for loops to make the code much shorter.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/40996

- Found files with invalid states:
  - DQM/SiTrackerPhase2/test/step3_pre4_inDQM.root:
    - Added: 4ab0c8b403c1928b4a8f20951a76a4741c3d4ebe
    - Deleted: 98f89e8f0404fdda07bafbeb96d25e98384033ad
  - DQM/SiTrackerPhase2/test/DQM_V0001_R000000001__Global__CMSSW_X_Y_Z__RECO.root:
    - Added: 4ab0c8b403c1928b4a8f20951a76a4741c3d4ebe
    - Deleted: 98f89e8f0404fdda07bafbeb96d25e98384033ad

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/40996/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/40996/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:
```diff
@@ -81,48 +128,279 @@ Phase2OTValidateTTStub::~Phase2OTValidateTTStub() {
+  theTrackerGeom_ = &(iSetup.getData(geomToken_));
+  tTopo_ = &(iSetup.getData(tTopoToken_));
+
+  // Clear existing histograms
```
Where does the idea to clear the existing histograms come from? It may be correct, but I don't see simiilar things happening in beginRun of the other DQM packages. e.g. https://github.com/brandiskip/cmssw/blob/test_branch/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateCluster.cc .

brandiskip replied:
I kept getting a seg fault and someone pointed me to lines 116 and 117 found here: https://cmssdt.cern.ch/lxr/source/DQM/HLTEvF/plugins/TriggerBxVsOrbitMonitor.cc?v=CMSSW_14_1_X_2024-07-17-2300
and when I followed that format, it fixed my issues. Perhaps there is a better way. I will investigate further.

tomalin replied:
You need to set the vector size to 6 or whatever, since you refer to the elements later on. e.g. https://github.com/brandiskip/cmssw/blob/test_branch/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc#L452 . But I don't understand why it needs resetting o 6 every run. Or why the call to clear is needed. If it were set to 6 in the constructor, resize would not be needed.

brandiskip replied:
Fixed

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:
```diff
@@ -81,48 +128,279 @@ Phase2OTValidateTTStub::~Phase2OTValidateTTStub() {
+        layer = -1;
+      }
+
+      int isPSmodule = (topo.nrows() == 960) ? 1 : 0;
```
Having magic numbers in CMSSW code, such as 960 is forbidden.

brandiskip replied:
Fixed

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:
```diff
@@ -50,29 +58,68 @@ class Phase2OTValidateTTStub : public DQMEDAnalyzer {
+  std::vector<MonitorElement*> bend_res_fw_endcap_discs;
+  std::vector<MonitorElement*> bend_res_bw_endcap_discs;
+
+  std::vector<MonitorElement*>* phi_res_discs = nullptr;
```
How can a vector equal a nullptr?

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:
```diff
@@ -81,48 +128,279 @@ Phase2OTValidateTTStub::~Phase2OTValidateTTStub() {
+  bend_res_bw_endcap_discs.clear();
+
+  // Resize vectors and set elements to nullptr
+  phi_res_barrel_layers.resize(6, nullptr);
```
Having magic numbers in the code like 5 or 6 is not allowed. At the very least these need to be named constants.

brandiskip replied:
Fixed

tomalin replied:
Has this really been fixed? I still see the magic numbers 5 & 6 in the code.

brandiskip replied:
Sorry, I missed those. This is now fixed.

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:
```diff
@@ -81,48 +128,279 @@ Phase2OTValidateTTStub::~Phase2OTValidateTTStub() {
+
+      // Ensure that the vectors are correctly assigned before use
+      if (isBarrel == 1) {
+          phi_res_discs = &phi_res_barrel_layers; // Pointing to barrel layers vector
```
What is going on here? If we're in the barrel, why is variable phi_res_discs being set?

brandiskip replied:
Fixed

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:
```diff
@@ -81,48 +128,279 @@ Phase2OTValidateTTStub::~Phase2OTValidateTTStub() {
+      }
+
+      // Fill the appropriate histogram based on layer/disc
+      if (layer >= 1 && layer <= 6) {
```
I can't understand what this part of hte code is doing. The comment suggests that its filling both layers and disks. But it requires layer <= 6, when there are only 5 disks. And it fills a variable alled bend_res_discs, which doesn't sound appropriate for the barrel.

brandiskip replied:
Fixed

tomalin commented:
A general question. How have you chosen the level of detail that is best for the validation code? e.g. Histograms for barrel + endcap vs histograms for every single layer?

tomalin commented:
General comments: All classes need a comment at the top explaining what they are for.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_14_2_X. Please open a backport if it should also go in to CMSSW_14_1_X.

antoniovilela commented:
ping (to make bot change milestone)

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/42588

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/42588/cms-checkout-topic.log

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_15_0_X. Please open a backport if it should also go in to CMSSW_14_2_X.

brandiskip committed:
Change input file and geometry to D98

brandiskip committed:
added code for residual histos

brandiskip committed:
added code for stub efficiency hists

brandiskip committed:
change folder naming scheme for efficiency and resolution hists

brandiskip committed:
added code for residual hists

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/42919

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/42919/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/42919/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

brandiskip committed:
revert to original

brandiskip committed:
changed input file and geometry to D98

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43058

cmsbuild commented:
Pull request #45475 was updated.

brandiskip committed:
added additional cuts

brandiskip committed:
added stub efficiency hists

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43185

cmsbuild commented:
Pull request #45475 was updated.

tomalin commented:
All cfg.py files should have a comment at the top explaining what they are for. So please add one to dqm_phase2trk_cfg.py & harvestingstep_phase2tk_cfg.py. Though the comments at the top of these two say they are "automatically generated", suggesting that if you add a comment, it would be deleted next time they are regenerated? If so, instead add a README.md to DQM/SiTrackerPhase2/ explaining how to run this code.

tomalin commented:
If I look at the line "process.schedule" dqmstep_phase2tk_cfg.py, in addition to running the code to make the DQM histograms, its also running "process.rechits_step" to remake the rechits. Why is it necessary to remake the rechits, rather than just using the ones in the input MC dataset? I note that it's not bothering to remake the TTClusters or TTStubs or to redo the association of the RecHits with the MC truth.

tomalin commented:
Using topo.nrows() == 960 to determine if a module is PS or 2S is ugly.  Can't you do something like https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-14_2_0_pre4/L1Trigger/TrackTrigger/src/SensorModule.cc#L51 ?

brandiskip commented:
> If I look at the line "process.schedule" dqmstep_phase2tk_cfg.py, in addition to running the code to make the DQM histograms, its also running "process.rechits_step" to remake the rechits. Why is it necessary to remake the rechits, rather than just using the ones in the input MC dataset? I note that it's not bothering to remake the TTClusters or TTStubs or to redo the association of the RecHits with the MC truth.

I'm not entirely sure why process.rechits_step is included. That part of the code was already like that in master CMSSW. The only reason that dqmstep_phase2tk_cfg.py is part of my PR is to fix the incorrect geometry configuration on [this line](https://github.com/cms-sw/cmssw/blob/master/DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py#L19). In the latest CMSSW release, that geometry doesn’t exist, so I corrected it, updated the input file, and made a corresponding change in harvestingstep_phase2tk_cfg.py.

brandiskip committed:
Add purpose and usage comments

brandiskip committed:
Replace hardcoded module check with getDetectorType for PS modules

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43289

cmsbuild commented:
Pull request #45475 was updated.

brandiskip committed:
Added README.md

cmsbuild commented:
Pull request #45475 was updated.

brandiskip commented:
> All cfg.py files should have a comment at the top explaining what they are for. So please add one to dqm_phase2trk_cfg.py & harvestingstep_phase2tk_cfg.py. Though the comments at the top of these two say they are "automatically generated", suggesting that if you add a comment, it would be deleted next time they are regenerated? If so, instead add a README.md to DQM/SiTrackerPhase2/ explaining how to run this code.

Done

brandiskip commented:
> Using topo.nrows() == 960 to determine if a module is PS or 2S is ugly. Can't you do something like https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-14_2_0_pre4/L1Trigger/TrackTrigger/src/SensorModule.cc#L51 ?

Done

brandiskip committed:
Add class-level comments to explain the purpose of each class

brandiskip commented:
> General comments: All classes need a comment at the top explaining what they are for.

Done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43353

cmsbuild commented:
Pull request #45475 was updated. @antoniovagnerini, @cmsbuild, @rseidita can you please check and sign again.

brandiskip commented:
> A general question. How have you chosen the level of detail that is best for the validation code? e.g. Histograms for barrel + endcap vs histograms for every single layer?

This is still being determined. As we learn more about how these histograms can best be used in release comparison (such as RelMon), we will know more about the best level of detail.

tomalin commented:
Are the names of the files Phase2OTHarvestTrackingParticles & Phase2ValidateTrackingParticles sensible? These names suggest that this code is used to validate the tracking particle collection, but the comments at the top suggest something very different.

tomalin commented on Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestTrackingParticles.cc:
```diff
@@ -38,65 +55,210 @@ void Phase2OTHarvestTrackingParticles::dqmEndJob(DQMStore::IBooker &ibooker, DQM
+    MonitorElement *merespt_eta2to2p4_pt8toInf =
+        dbe->get(topFolderName_ + "/ResolutionIngredients/respt_eta2to2p4_pt8toInf");
+
+    MonitorElement *mereseta_eta0to0p7 = dbe->get(topFolderName_ + "/ResolutionIngredients/reseta_eta0to0p7");
```
On a related note, although the code at https://github.com/cms-sw/cmssw/blob/master/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTrackingParticles.cc#L427 wasn't added by this PR, it is another example of very long-winded, repetitive code, that could probably be massively shortened using vectors and loops. This stuff really needs a huge clean up.

brandiskip replied:
Since the code in question was not introduced by this PR, I propose that we proceed with merging the current updates and address the cleanup in a follow-up PR. This cleanup can potentially be done in tandem with adding histograms for displaced tracking.

brandiskip commented:
> Are the names of the files Phase2OTHarvestTrackingParticles & Phase2ValidateTrackingParticles sensible? These names suggest that this code is used to validate the tracking particle collection, but the comments at the top suggest something very different.

Right, good point. The Phase2ValidateTrackingParticles loops over TPs which are used as the denominator of the efficiency. So basically, this code validates how well reconstructed objects (like stubs and tracks) match tracking particles. So perhaps something like Phase2ValidateReconstruction would be better?

tomalin commented:
> > Are the names of the files Phase2OTHarvestTrackingParticles & Phase2ValidateTrackingParticles sensible? These names suggest that this code is used to validate the tracking particle collection, but the comments at the top suggest something very different.
> 
> Right, good point. The Phase2ValidateTrackingParticles loops over TPs which are used as the denominator of the efficiency. So basically, this code validates how well reconstructed objects (like stubs and tracks) match tracking particles. So perhaps something like Phase2ValidateReconstruction would be better?

You understand what these classes do better than me, so I leave it to you. I haven't understood why we even need the harvesting step. It seems a bit of a nightmare in the current code, since the harvesting code must assume which histograms were created by the validate code, but the list of histograms is not shared between the two bits of code. Hence anyone changing a histogram in the validation code is likely to break the harvesting code. If true, then this again makes the code a nightmare to maintain. If harvesting is only needed to make efficiency histograms, why can't these be produced directly in the validation step using ROOT's TEfficiency class? How do other subdetectors do this?

tomalin commented:
A few of Suchandra's comments don't seem to be answered? e.g. https://github.com/cms-sw/cmssw/pull/45475#discussion_r1680391016 .

brandiskip commented:
> A few of Suvankar's comments don't seem to be answered? e.g. [#45475 (comment)](https://github.com/cms-sw/cmssw/pull/45475#discussion_r1680391016) .

Sorry, I somehow missed these. I will review and address them.

tomalin commented:
> > If harvesting is only needed to make efficiency histograms, why can't these be produced directly in the validation step using ROOT's TEfficiency class? How do other subdetectors do this?
> 
> good luck with merging different `TEfficiency` plots across an arbitrary number of DQMIO files :)

Hi @mmusich , if TEfficiency is not a good solution, can you recommend any beautiful harvesting code that makes efficiency histograms that we could use as an example to follow? I'd like a solution where the harvesting code doesn't need to make hard-wired assumptions about which histograms the validation code produced, particularly when there are dozens of histograms for different Pt or eta intervals.

mmusich commented:
>  I'd like a solution where the harvesting code doesn't need to make hard-wired assumptions about which histograms the validation code produced, particularly when there are dozens of histograms for different Pt or eta intervals.

few ideas:

- store the list of histogram booked in the validation step somewhere that you can access from the harvesting step (example that comes to mind [SiPixelLorentzAngleCalibrationStruct](https://github.com/cms-sw/cmssw/blob/75451d59a7acc30aec874be9a6b9a8835f2f7b3e/CalibTracker/SiPixelLorentzAngle/interface/SiPixelLorentzAngleCalibrationStruct.h) that is used by both the "[DQM](https://github.com/cms-sw/cmssw/blob/75451d59a7acc30aec874be9a6b9a8835f2f7b3e/CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc)" and the "[Havesting](https://github.com/cms-sw/cmssw/blob/75451d59a7acc30aec874be9a6b9a8835f2f7b3e/CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLHarvester.cc)" codes). 
- make the harvester code generic enough to decide at run time, histogram by histogram what it needs to do depending on some regexp on the name.

as for examples, the whole cmssw is structured more or less  in this way. Perhaps, looking into https://github.com/cms-sw/cmssw/tree/master/DQM/TrackingMonitorClient might be a starting point (not guaranteeing it's the best one -- I am not a DQM expert, I would let @cms-sw/dqm-l2 give more suggestions).

brandiskip commented on Validation/SiTrackerPhase2V/python/Phase2OTValidateTTStub_cfi.py:
```diff
@@ -0,0 +1,20 @@
+import FWCore.ParameterSet.Config as cms
```
Resolved: This file has been removed

brandiskip commented on Validation/SiTrackerPhase2V/python/Phase2OTValidateTrackingParticles_cfi.py:
```diff
@@ -0,0 +1,134 @@
+import FWCore.ParameterSet.Config as cms
```
Resolved: This file has been removed

mmusich commented:
>  so avoiding the need for a harvesting step -- TEfficiency objects have an add() function just like TH1 objects.

is this interface thread-friendly?

> It should therefore easy to add separate .root files containing TEfficiency objects together. 

the cmssw DQM framework doesn't support `TEfficiency` as a bookable `MonitorElement` underlying type,

https://github.com/cms-sw/cmssw/blob/79a21738b3213206b3e20f8588ef3f3d455b4f62/DQMServices/Core/interface/DQMNet.h#L27-L46

Of course, you're free to try to implement support for it, following the general requirement and specifications for cmssw development.

brandiskip committed:
renamed to better reflect functionality

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43435

ERROR: Build errors found during clang-tidy run.
```
src/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:235:20: error: no member named 'reduceRange' in namespace 'reco'; did you mean 'reducePhiRange'? [clang-diagnostic-error]
  235 |     tp_phi = reco::reduceRange(tp_phi);
      |              ~~~~~~^~~~~~~~~~~
      |                    reducePhiRange
--
src/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:240:20: error: no member named 'reduceRange' in namespace 'reco'; did you mean 'reducePhiRange'? [clang-diagnostic-error]
  240 |     tp_phi = reco::reduceRange(tp_phi);
      |              ~~~~~~^~~~~~~~~~~
      |                    reducePhiRange
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

brandiskip committed:
re-ran code-checks and code-format

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43436

ERROR: Build errors found during clang-tidy run.
```
src/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:236:20: error: no member named 'reduceRange' in namespace 'reco'; did you mean 'reducePhiRange'? [clang-diagnostic-error]
  236 |     tp_phi = reco::reduceRange(tp_phi);
      |              ~~~~~~^~~~~~~~~~~
      |                    reducePhiRange
--
src/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:241:20: error: no member named 'reduceRange' in namespace 'reco'; did you mean 'reducePhiRange'? [clang-diagnostic-error]
  241 |     tp_phi = reco::reduceRange(tp_phi);
      |              ~~~~~~^~~~~~~~~~~
      |                    reducePhiRange
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

brandiskip committed:
made edits required for new file names Phase2OTHarvestReconstruction.cc and Phase2OTValidateReconstruction.cc

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43456

ERROR: Build errors found during clang-tidy run.
```
src/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:235:20: error: no member named 'reduceRange' in namespace 'reco'; did you mean 'reducePhiRange'? [clang-diagnostic-error]
  235 |     tp_phi = reco::reduceRange(tp_phi);
      |              ~~~~~~^~~~~~~~~~~
      |                    reducePhiRange
--
src/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateTTStub.cc:240:20: error: no member named 'reduceRange' in namespace 'reco'; did you mean 'reducePhiRange'? [clang-diagnostic-error]
  240 |     tp_phi = reco::reduceRange(tp_phi);
      |              ~~~~~~^~~~~~~~~~~
      |                    reducePhiRange
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

brandiskip committed:
change reduceRange to reducePhiRange

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43457

cmsbuild commented:
Pull request #45475 was updated. @antoniovagnerini, @cmsbuild, @rseidita can you please check and sign again.

brandiskip commented:
> Are the names of the files Phase2OTHarvestTrackingParticles & Phase2ValidateTrackingParticles sensible? These names suggest that this code is used to validate the tracking particle collection, but the comments at the top suggest something very different.

Names have been changed to Phase2OTHarvestReconstruction and Phase2OTValidateReconstruction.

mmusich commented on Validation/SiTrackerPhase2V/plugins/Phase2OTHarvestReconstruction.cc:
```diff
@@ -38,65 +55,210 @@ void Phase2OTHarvestTrackingParticles::dqmEndJob(DQMStore::IBooker &ibooker, DQM
-    MonitorElement *meresd0_eta1p2to1p6 = dbe->get(topFolderName_ + "/Resolution/resd0_eta1p2to1p6");
-    MonitorElement *meresd0_eta1p6to2 = dbe->get(topFolderName_ + "/Resolution/resd0_eta1p6to2");
-    MonitorElement *meresd0_eta2to2p4 = dbe->get(topFolderName_ + "/Resolution/resd0_eta2to2p4");
+    MonitorElement *meN_clus_barrel = dbe->get(topFolderName_ + "/EfficiencyIngredients/gen_clusters_if_stub_barrel");
```
perhaps one could profit of this PR to get rid of the deprecated access pattern through `edm::Service<DQMStore>().operator->()` in favor of using `DQMStore::IGetter`.

brandiskip replied:
done

brandiskip committed:
Replace deprecated edm::Service<DQMStore>() access with DQMStore::IGetter

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43473

cmsbuild commented:
Pull request #45475 was updated. @antoniovagnerini, @cmsbuild, @rseidita can you please check and sign again.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_15_1_X. Please open a backport if it should also go in to CMSSW_15_0_X.

brandiskip committed:
Replaced magic numbers with N_LAYER and N_DISK

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45475/43617

cmsbuild commented:
Pull request #45475 was updated. @antoniovagnerini, @cmsbuild, @rseidita can you please check and sign again.

brandiskip commented:
Can this PR be merged?

rseidita commented:
please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e76ee2/44442/summary.html
**COMMIT**: e5cf02822bd93b27ae50a5eda058b34943aa6f70
**CMSSW**: CMSSW_15_1_X_2025-02-17-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45475/44442/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 4 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-17-1100+e76ee2/66958/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 6 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-17-1100+e76ee2/66958/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3905090
 - DQMHistoTests: [Total failures: 55](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-17-1100+e76ee2/66958/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3905015
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 352.296 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 24834.911,... ): 55.916 KiB       TrackerPhase2OTStubV/Residual
 - DQMHistoSizes: changed ( 24834.911,... ): 21.150 KiB       TrackerPhase2OTL1TrackV/ResolutionIngredients
 - DQMHistoSizes: changed ( 24834.911,... ): -20.667 KiB      TrackerPhase2OTL1TrackV/Resolution
 - DQMHistoSizes: changed ( 24834.911,... ): 5.277 KiB        TrackerPhase2OTL1TrackV/EfficiencyIngredients
 - DQMHistoSizes: changed ( 24834.911,... ): -2.961 KiB       TrackerPhase2OTL1TrackV/Efficiency
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-17-1100+e76ee2/66958/triggerResults): no differences found

rseidita commented on DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py:
```diff
@@ -16,7 +23,7 @@
 process.load('Configuration.EventContent.EventContent_cff')
 process.load('SimGeneral.MixingModule.mixNoPU_cfi')
-process.load('Configuration.Geometry.GeometryExtendedRun4D76Reco_cff')
+process.load('Configuration.Geometry.GeometryExtendedRun4D98Reco_cff')
```
Small differences in some distributions (e.g., [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-17-1100+e76ee2/66958/29634.0_TTbar_14TeV+Run4D110/TrackerPhase2OTL1TrackV__trackParticles_trackParts_Pt.png)) in WFs 29634.0 and 29634.911. Is this due to the change in geometry from D76 to D98?

mmusich replied:
this is a test configuration, thus it should have no bearing on the DQM results of centrally run workflows.

brandiskip replied:
In Phase2OTValidateReconstruction, in line 1425, I changed TP_minPt from 2.0 to 1.5 and that's causing the trackParts_Pt hist to be slightly different.

rseidita commented on Validation/SiTrackerPhase2V/python/Phase2ValidateL1TTObjects_cff.py:
```diff
@@ -1,8 +1,8 @@
 import FWCore.ParameterSet.Config as cms
 
-from Validation.SiTrackerPhase2V.Phase2OTValidateTrackingParticles_cfi import * 
+from Validation.SiTrackerPhase2V.Phase2OTValidateReconstruction_cfi import * 
```
@brandiskip could you point me to this file? As far as I can see it's not in the codebased up to 15_1_X_2025_02_17_2300 and it's not in this PR?

mmusich replied:
https://github.com/cms-sw/cmssw/blob/e5cf02822bd93b27ae50a5eda058b34943aa6f70/Validation/SiTrackerPhase2V/plugins/Phase2OTValidateReconstruction.cc#L1307-L1431

rseidita commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @mandrenguyen, @rappoccio, @antoniovilela, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)
